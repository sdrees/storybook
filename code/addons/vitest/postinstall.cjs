module.exports = require('./dist/postinstall.js');
