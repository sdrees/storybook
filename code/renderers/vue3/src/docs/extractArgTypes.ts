import type { SBType, StrictArgTypes } from '@storybook/types';

import {
  extractComponentProps,
  convert as genericConvert,
  hasDocgen,
  type ArgTypesExtractor,
  type DocgenInfo,
} from '@storybook/docs-tools';

type Schema = { kind: string; schema: [] | object; type?: string } | string;
type MetaDocgenInfo = DocgenInfo & {
  type: string | { name: string; value: string[] };
  default: string;
  global: boolean;
  name: string;
  schema?: Schema;
  tags?: { name: string; text: string }[];
};

// "exposed" is used by the vue-component-meta plugin while "expose" is used by vue-docgen-api
const ARG_TYPE_SECTIONS = ['props', 'events', 'slots', 'exposed', 'expose'] as const;

export const extractArgTypes: ArgTypesExtractor = (component) => {
  if (!hasDocgen(component)) {
    return null;
  }

  const argTypes: StrictArgTypes = {};

  ARG_TYPE_SECTIONS.forEach((section) => {
    const props = extractComponentProps(component, section);

    props.forEach((extractedProp) => {
      const docgenInfo = extractedProp.docgenInfo as MetaDocgenInfo;

      if (argTypes[docgenInfo.name] || docgenInfo.global) {
        return; // skip duplicate and global props
      }

      const type =
        typeof docgenInfo.type === 'string' ? docgenInfo.type : docgenInfo.type?.name ?? '';
      const sbType = section === 'props' ? convertPropType(docgenInfo) : ({ name: type } as SBType);

      const defaultValue = { summary: docgenInfo.default };

      argTypes[docgenInfo.name] = {
        name: docgenInfo.name,
        description: formatDescriptionWithTags(docgenInfo.description, docgenInfo.tags),
        defaultValue,
        type: sbType,
        table: {
          type: { summary: type.replace(' | undefined', '') },
          jsDocTags: docgenInfo.tags ?? [],
          defaultValue,
          category: section,
        },
        control: { disabled: !['props', 'slots'].includes(section) },
      };
    });
  });

  return argTypes;
};

/**
 * Converts the given prop type into a SBType so it can be correctly displayed in the UI (controls etc.).
 */
export const convertPropType = (propInfo: MetaDocgenInfo): SBType => {
  const schema = propInfo.schema;
  const required = propInfo.required ?? false;
  const fallbackSbType = { name: schema, required } as SBType;

  if (!schema) return genericConvert(propInfo) ?? fallbackSbType;
  if (typeof schema === 'string') return fallbackSbType;

  // convert enum schemas (e.g. union or enum type to corresponding SBType)
  //  so the enum values can be selected via radio/dropdown in the UI
  if (schema.kind === 'enum' && Array.isArray(schema.schema)) {
    // filter out empty or "undefined" for optional props
    const definedValues = schema.schema.filter((item) => item != null && item !== 'undefined');

    if (definedValues.length === 1 && typeof definedValues[0] === 'object') {
      return convertPropType({ schema: definedValues[0] } as MetaDocgenInfo);
    }

    const values = definedValues
      .filter((item: Schema) => typeof item === 'string')
      .map((item: Schema) => (typeof item !== 'string' ? item.schema.toString() : item))
      .map((item: string) => item.replace(/'/g, '"'));

    if (values.length === 0) return fallbackSbType;

    const isBoolean = values.length === 2 && values.includes('true') && values.includes('false');
    if (isBoolean) return { name: 'boolean', required };

    const isLateralUnion =
      values.length > 1 && values.every((item) => item.startsWith('"') && item.endsWith('"'));
    const isEnum =
      !isLateralUnion &&
      values.length > 1 &&
      values.every((item) => typeof item === 'string' && item.includes('.'));

    if (isLateralUnion || isEnum) {
      const valuesWithoutQuotes = values.map((item: string) => item.replace(/"/g, ''));
      return { name: 'enum', value: valuesWithoutQuotes, required };
    }

    return { name: values[0], required } as SBType;
  }

  // recursively convert object properties to SBType
  if (schema.kind === 'object' && typeof schema.schema === 'object') {
    const schemaObject = schema.schema as Record<string, MetaDocgenInfo>;

    return {
      name: 'object',
      required,
      value: Object.entries(schemaObject).reduce<Record<string, SBType>>((obj, [key, value]) => {
        obj[key] = convertPropType(value);
        return obj;
      }, {}),
    };
  }

  if (schema.kind === 'array' && Array.isArray(schema.schema)) {
    return {
      name: 'array',
      value: convertPropType({ schema: schema.schema[0] } as MetaDocgenInfo),
      required,
    };
  }

  return fallbackSbType;
};

/**
 * Adds the descriptions for the given tags if available.
 */
const formatDescriptionWithTags = (description: string, tags: MetaDocgenInfo['tags']): string => {
  if (!tags?.length || !description) return description ?? '';
  const tagDescriptions = tags.map((tag) => `@${tag.name}: ${tag.text}`).join('<br>');
  return `${tagDescriptions}<br><br>${description}`;
};
