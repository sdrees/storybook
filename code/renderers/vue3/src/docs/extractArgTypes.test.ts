import { extractComponentProps, hasDocgen } from '@storybook/docs-tools';
import type { Mock } from 'vitest';
import { beforeEach, describe, expect, it, vi, vitest } from 'vitest';
import { extractArgTypes } from './extractArgTypes';
import {
  mockExtractComponentEventsReturn,
  mockExtractComponentPropsReturn,
  mockExtractComponentSlotsReturn,
  referenceTypeEvents,
  referenceTypeProps,
  templateSlots,
} from './tests-meta-components/meta-components';

vitest.mock('@storybook/docs-tools');

describe('extractArgTypes', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  it('should return null if component does not contain docs', () => {
    (hasDocgen as Mock).mockReturnValueOnce(false);
    (extractComponentProps as Mock).mockReturnValueOnce([] as any);

    expect(extractArgTypes({} as any)).toBeNull();
  });

  it('should extract arg types for component', () => {
    const component = referenceTypeProps;
    (hasDocgen as Mock).mockReturnValueOnce(true);
    (extractComponentProps as Mock).mockReturnValue(mockExtractComponentPropsReturn);

    const argTypes = extractArgTypes(component);

    expect(argTypes).toMatchInlineSnapshot(`
      {
        "array": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description required array object",
          "name": "array",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedProps[]",
            },
          },
          "type": {
            "name": "array",
            "required": true,
            "value": {
              "name": "object",
              "required": false,
              "value": {
                "nestedProp": {
                  "name": "string",
                  "required": true,
                },
              },
            },
          },
        },
        "arrayOptional": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description optional array object",
          "name": "arrayOptional",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedProps[]",
            },
          },
          "type": {
            "name": "array",
            "required": false,
            "value": {
              "name": "object",
              "required": false,
              "value": {
                "nestedProp": {
                  "name": "string",
                  "required": true,
                },
              },
            },
          },
        },
        "bar": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": "1",
          },
          "description": "description bar is optional number",
          "name": "bar",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": "1",
            },
            "jsDocTags": [],
            "type": {
              "summary": "number",
            },
          },
          "type": {
            "name": "number",
            "required": false,
          },
        },
        "baz": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description baz is required boolean",
          "name": "baz",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "boolean",
            },
          },
          "type": {
            "name": "boolean",
            "required": true,
          },
        },
        "enumValue": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description enum value",
          "name": "enumValue",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyEnum",
            },
          },
          "type": {
            "name": "enum",
            "required": true,
            "value": [
              "MyEnum.Small",
              "MyEnum.Medium",
              "MyEnum.Large",
            ],
          },
        },
        "foo": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "@default: "rounded"<br>@since: v1.0.0<br>@see: https://vuejs.org/<br>@deprecated: v1.1.0<br><br>string foo",
          "name": "foo",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [
              {
                "name": "default",
                "text": ""rounded"",
              },
              {
                "name": "since",
                "text": "v1.0.0",
              },
              {
                "name": "see",
                "text": "https://vuejs.org/",
              },
              {
                "name": "deprecated",
                "text": "v1.1.0",
              },
            ],
            "type": {
              "summary": "string",
            },
          },
          "type": {
            "name": "string",
            "required": true,
          },
        },
        "inlined": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "",
          "name": "inlined",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "{ foo: string; }",
            },
          },
          "type": {
            "name": "object",
            "required": true,
            "value": {
              "foo": {
                "name": "string",
                "required": true,
              },
            },
          },
        },
        "literalFromContext": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description literal type alias that require context",
          "name": "literalFromContext",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": ""Uncategorized" | "Content" | "Interaction" | "Display" | "Forms" | "Addons"",
            },
          },
          "type": {
            "name": "enum",
            "required": true,
            "value": [
              "Uncategorized",
              "Content",
              "Interaction",
              "Display",
              "Forms",
              "Addons",
            ],
          },
        },
        "nested": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description nested is required nested object",
          "name": "nested",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedProps",
            },
          },
          "type": {
            "name": "object",
            "required": true,
            "value": {
              "nestedProp": {
                "name": "string",
                "required": true,
              },
            },
          },
        },
        "nestedIntersection": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description required nested object with intersection",
          "name": "nestedIntersection",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedProps & { additionalProp: string; }",
            },
          },
          "type": {
            "name": "object",
            "required": true,
            "value": {
              "additionalProp": {
                "name": "string",
                "required": true,
              },
              "nestedProp": {
                "name": "string",
                "required": true,
              },
            },
          },
        },
        "nestedOptional": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description optional nested object",
          "name": "nestedOptional",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedProps | MyIgnoredNestedProps",
            },
          },
          "type": {
            "name": {
              "kind": "enum",
              "schema": [
                "undefined",
                {
                  "kind": "object",
                  "schema": {
                    "nestedProp": {
                      "declarations": [],
                      "description": "nested prop documentation",
                      "global": false,
                      "name": "nestedProp",
                      "required": true,
                      "schema": "string",
                      "tags": [],
                      "type": "string",
                    },
                  },
                  "type": "MyNestedProps",
                },
                {
                  "kind": "object",
                  "schema": {
                    "nestedProp": {
                      "declarations": [],
                      "description": "",
                      "global": false,
                      "name": "nestedProp",
                      "required": true,
                      "schema": "string",
                      "tags": [],
                      "type": "string",
                    },
                  },
                  "type": "MyIgnoredNestedProps",
                },
              ],
              "type": "MyNestedProps | MyIgnoredNestedProps | undefined",
            },
            "required": false,
          },
        },
        "recursive": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "",
          "name": "recursive",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "MyNestedRecursiveProps",
            },
          },
          "type": {
            "name": "object",
            "required": false,
            "value": {
              "recursive": {
                "name": "MyNestedRecursiveProps",
                "required": true,
              },
            },
          },
        },
        "stringArray": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": "["foo", "bar"]",
          },
          "description": "description stringArray is string array",
          "name": "stringArray",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": "["foo", "bar"]",
            },
            "jsDocTags": [],
            "type": {
              "summary": "string[]",
            },
          },
          "type": {
            "name": "array",
            "required": false,
            "value": {
              "name": "string",
              "required": false,
            },
          },
        },
        "union": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description union is required union type",
          "name": "union",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "string | number",
            },
          },
          "type": {
            "name": "string",
            "required": true,
          },
        },
        "unionOptional": {
          "control": {
            "disabled": false,
          },
          "defaultValue": {
            "summary": undefined,
          },
          "description": "description unionOptional is optional union type",
          "name": "unionOptional",
          "table": {
            "category": "props",
            "defaultValue": {
              "summary": undefined,
            },
            "jsDocTags": [],
            "type": {
              "summary": "string | number | boolean",
            },
          },
          "type": {
            "name": "string",
            "required": false,
          },
        },
      }
    `);
  });

  it('should extract events for Vue component', () => {
    const component = referenceTypeEvents;
    (hasDocgen as Mock).mockReturnValueOnce(true);
    (extractComponentProps as Mock).mockReturnValue(mockExtractComponentEventsReturn);

    const argTypes = extractArgTypes(component);

    expect(argTypes).toMatchSnapshot();
  });

  it('should extract slots type for Vue component', () => {
    const component = templateSlots;
    (hasDocgen as Mock).mockReturnValueOnce(true);
    (extractComponentProps as Mock).mockReturnValue(mockExtractComponentSlotsReturn);

    const argTypes = extractArgTypes(component);

    expect(argTypes).toMatchSnapshot();
  });
});
