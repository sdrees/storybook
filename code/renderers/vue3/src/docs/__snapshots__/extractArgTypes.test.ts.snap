// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractArgTypes (vue-docgen-api) > should extract events for Vue component 1`] = `
{
  "bar": {
    "control": {
      "disabled": true,
    },
    "description": "",
    "name": "bar",
    "table": {
      "category": "events",
      "type": {
        "summary": "[value: { year: number; title?: any; }]",
      },
    },
    "type": {
      "name": "other",
      "value": "[value: { year: number; title?: any; }]",
    },
  },
  "baz": {
    "control": {
      "disabled": true,
    },
    "description": "",
    "name": "baz",
    "table": {
      "category": "events",
      "type": {
        "summary": "[]",
      },
    },
    "type": {
      "name": "other",
      "value": "[]",
    },
  },
  "foo": {
    "control": {
      "disabled": true,
    },
    "description": "",
    "name": "foo",
    "table": {
      "category": "events",
      "type": {
        "summary": "[data?: { foo: string; }]",
      },
    },
    "type": {
      "name": "other",
      "value": "[data?: { foo: string; } | undefined]",
    },
  },
}
`;

exports[`extractArgTypes (vue-docgen-api) > should extract events for component 1`] = `
{
  "bar": {
    "control": {
      "disabled": true,
    },
    "description": "Test description bar",
    "name": "bar",
    "table": {
      "category": "events",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "{ year: number; title?: any }",
      },
    },
    "type": {
      "name": "other",
      "required": false,
      "value": "{ year: number; title?: any }",
    },
  },
  "baz": {
    "control": {
      "disabled": true,
    },
    "description": "Test description baz",
    "name": "baz",
    "table": {
      "category": "events",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": undefined,
    },
    "type": {
      "name": "other",
      "required": false,
      "value": "",
    },
  },
}
`;

exports[`extractArgTypes (vue-docgen-api) > should extract expose for component 1`] = `
{
  "count": {
    "control": {
      "disabled": true,
    },
    "description": "a count number",
    "name": "count",
    "table": {
      "category": "expose",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": undefined,
    },
    "type": {
      "name": "other",
      "value": "",
    },
  },
  "label": {
    "control": {
      "disabled": true,
    },
    "description": "a label string",
    "name": "label",
    "table": {
      "category": "expose",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": undefined,
    },
    "type": {
      "name": "other",
      "value": "",
    },
  },
}
`;

exports[`extractArgTypes (vue-docgen-api) > should extract props for component 1`] = `
{
  "array": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "description required array object",
    "name": "array",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "MyNestedProps[]",
      },
    },
    "type": {
      "name": "array",
      "required": true,
      "value": {
        "name": "object",
        "required": false,
        "value": {
          "nestedProp": {
            "name": "string",
            "required": true,
          },
        },
      },
    },
  },
  "arrayOptional": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "description optional array object",
    "name": "arrayOptional",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "MyNestedProps[]",
      },
    },
    "type": {
      "name": "array",
      "required": false,
      "value": {
        "name": "object",
        "required": false,
        "value": {
          "nestedProp": {
            "name": "string",
            "required": true,
          },
        },
      },
    },
  },
  "bar": {
    "control": {
      "disabled": false,
    },
    "defaultValue": {
      "summary": "1",
    },
    "description": "description bar is optional number",
    "name": "bar",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": "1",
      },
      "type": {
        "summary": "number",
      },
    },
    "type": {
      "name": "number",
      "required": false,
    },
  },
  "baz": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "description baz is required boolean",
    "name": "baz",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "boolean",
      },
    },
    "type": {
      "name": "boolean",
      "required": true,
    },
  },
  "enumValue": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "description enum value",
    "name": "enumValue",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "MyEnum",
      },
    },
    "type": {
      "name": "enum",
      "required": true,
      "value": [
        "MyEnum.Small",
        "MyEnum.Medium",
        "MyEnum.Large",
      ],
    },
  },
  "foo": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "@default: "rounded"<br>@since: v1.0.0<br>@see: https://vuejs.org/<br>@deprecated: v1.1.0<br><br>string foo",
    "name": "foo",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "string",
      },
    },
    "type": {
      "name": "string",
      "required": true,
    },
  },
  "inlined": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "",
    "name": "inlined",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "{ foo: string; }",
      },
    },
    "type": {
      "name": "object",
      "required": true,
      "value": {
        "foo": {
          "name": "string",
          "required": true,
        },
      },
    },
  },
  "literalFromContext": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "description literal type alias that require context",
    "name": "literalFromContext",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": ""Uncategorized" | "Content" | "Interaction" | "Display" | "Forms" | "Addons"",
      },
    },
    "type": {
      "name": "enum",
      "required": true,
      "value": [
        "Uncategorized",
        "Content",
        "Interaction",
        "Display",
        "Forms",
        "Addons",
      ],
    },
  },
  "nested": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "description nested is required nested object",
    "name": "nested",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "MyNestedProps",
      },
    },
    "type": {
      "name": "object",
      "required": true,
      "value": {
        "nestedProp": {
          "name": "string",
          "required": true,
        },
      },
    },
  },
  "nestedIntersection": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "description required nested object with intersection",
    "name": "nestedIntersection",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "MyNestedProps & { additionalProp: string; }",
      },
    },
    "type": {
      "name": "object",
      "required": true,
      "value": {
        "additionalProp": {
          "name": "string",
          "required": true,
        },
        "nestedProp": {
          "name": "string",
          "required": true,
        },
      },
    },
  },
  "nestedOptional": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "description optional nested object",
    "name": "nestedOptional",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "MyNestedProps | MyIgnoredNestedProps",
      },
    },
    "type": {
      "name": "union",
      "required": false,
      "value": [
        {
          "name": "object",
          "required": false,
          "value": {
            "nestedProp": {
              "name": "string",
              "required": true,
            },
          },
        },
        {
          "name": "object",
          "required": false,
          "value": {
            "nestedProp": {
              "name": "string",
              "required": true,
            },
          },
        },
      ],
    },
  },
  "recursive": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "",
    "name": "recursive",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "MyNestedRecursiveProps",
      },
    },
    "type": {
      "name": "object",
      "required": false,
      "value": {
        "recursive": {
          "name": "other",
          "required": true,
          "value": "MyNestedRecursiveProps",
        },
      },
    },
  },
  "stringArray": {
    "control": {
      "disabled": false,
    },
    "defaultValue": {
      "summary": "["foo", "bar"]",
    },
    "description": "description stringArray is string array",
    "name": "stringArray",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": "["foo", "bar"]",
      },
      "type": {
        "summary": "string[]",
      },
    },
    "type": {
      "name": "array",
      "required": false,
      "value": {
        "name": "string",
        "required": false,
      },
    },
  },
  "union": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "description union is required union type",
    "name": "union",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "string | number",
      },
    },
    "type": {
      "name": "union",
      "required": true,
      "value": [
        {
          "name": "string",
          "required": false,
        },
        {
          "name": "number",
          "required": false,
        },
      ],
    },
  },
  "unionOptional": {
    "control": {
      "disabled": false,
    },
    "defaultValue": undefined,
    "description": "description unionOptional is optional union type",
    "name": "unionOptional",
    "table": {
      "category": "props",
      "defaultValue": undefined,
      "type": {
        "summary": "string | number | boolean",
      },
    },
    "type": {
      "name": "union",
      "required": false,
      "value": [
        {
          "name": "string",
          "required": false,
        },
        {
          "name": "number",
          "required": false,
        },
        {
          "name": "boolean",
          "required": false,
        },
      ],
    },
  },
}
`;

exports[`extractArgTypes (vue-docgen-api) > should extract props for component 2`] = `
{
  "array": {
    "control": {
      "disabled": false,
    },
    "description": "description required array object",
    "name": "array",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "MyNestedProps[]",
      },
    },
    "type": {
      "name": "other",
      "required": true,
      "value": "Array([object Object])",
    },
  },
  "arrayOptional": {
    "control": {
      "disabled": false,
    },
    "description": "description optional array object",
    "name": "arrayOptional",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "MyNestedProps[]",
      },
    },
    "type": {
      "name": "other",
      "required": false,
      "value": "Array([object Object])",
    },
  },
  "bar": {
    "control": {
      "disabled": false,
    },
    "description": "description bar is optional number",
    "name": "bar",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": "1",
      },
      "jsDocTags": undefined,
      "type": {
        "summary": "number",
      },
    },
    "type": {
      "name": "number",
      "required": false,
    },
  },
  "baz": {
    "control": {
      "disabled": false,
    },
    "description": "description baz is required boolean",
    "name": "baz",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "boolean",
      },
    },
    "type": {
      "name": "boolean",
      "required": true,
    },
  },
  "enumValue": {
    "control": {
      "disabled": false,
    },
    "description": "description enum value",
    "name": "enumValue",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "MyEnum",
      },
    },
    "type": {
      "name": "other",
      "required": true,
      "value": "MyEnum",
    },
  },
  "foo": {
    "control": {
      "disabled": false,
    },
    "description": "string foo",
    "name": "foo",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "string",
      },
    },
    "type": {
      "name": "string",
      "required": true,
    },
  },
  "inlined": {
    "control": {
      "disabled": false,
    },
    "description": undefined,
    "name": "inlined",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "{ foo: string }",
      },
    },
    "type": {
      "name": "other",
      "required": true,
      "value": "{ foo: string }",
    },
  },
  "literalFromContext": {
    "control": {
      "disabled": false,
    },
    "description": "description literal type alias that require context",
    "name": "literalFromContext",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "MyCategories",
      },
    },
    "type": {
      "name": "other",
      "required": true,
      "value": "MyCategories",
    },
  },
  "nested": {
    "control": {
      "disabled": false,
    },
    "description": "description nested is required nested object",
    "name": "nested",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "MyNestedProps",
      },
    },
    "type": {
      "name": "other",
      "required": true,
      "value": "MyNestedProps",
    },
  },
  "nestedIntersection": {
    "control": {
      "disabled": false,
    },
    "description": "description required nested object with intersection",
    "name": "nestedIntersection",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "MyNestedProps & {
  /**
   * description required additional property
   */
  additionalProp: string;
}",
      },
    },
    "type": {
      "name": "other",
      "required": true,
      "value": "intersection([object Object],[object Object])",
    },
  },
  "nestedOptional": {
    "control": {
      "disabled": false,
    },
    "description": "description optional nested object",
    "name": "nestedOptional",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "MyNestedProps | MyIgnoredNestedProps",
      },
    },
    "type": {
      "name": "union",
      "required": false,
      "value": [
        {
          "name": "other",
          "value": "MyNestedProps",
        },
        {
          "name": "other",
          "value": "MyIgnoredNestedProps",
        },
      ],
    },
  },
  "recursive": {
    "control": {
      "disabled": false,
    },
    "description": undefined,
    "name": "recursive",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "MyNestedRecursiveProps",
      },
    },
    "type": {
      "name": "other",
      "required": false,
      "value": "MyNestedRecursiveProps",
    },
  },
  "stringArray": {
    "control": {
      "disabled": false,
    },
    "description": "description stringArray is string array",
    "name": "stringArray",
    "table": {
      "category": "props",
      "defaultValue": {
        "summary": "() => ['foo', 'bar']",
      },
      "jsDocTags": undefined,
      "type": {
        "summary": "string[]",
      },
    },
    "type": {
      "name": "other",
      "required": false,
      "value": "Array([object Object])",
    },
  },
  "union": {
    "control": {
      "disabled": false,
    },
    "description": "description union is required union type",
    "name": "union",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "string | number",
      },
    },
    "type": {
      "name": "union",
      "required": true,
      "value": [
        {
          "name": "string",
        },
        {
          "name": "number",
        },
      ],
    },
  },
  "unionOptional": {
    "control": {
      "disabled": false,
    },
    "description": "description unionOptional is optional union type",
    "name": "unionOptional",
    "table": {
      "category": "props",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "string | number | boolean",
      },
    },
    "type": {
      "name": "union",
      "required": false,
      "value": [
        {
          "name": "string",
        },
        {
          "name": "number",
        },
        {
          "name": "boolean",
        },
      ],
    },
  },
}
`;

exports[`extractArgTypes (vue-docgen-api) > should extract slots for component 1`] = `
{
  "default": {
    "control": {
      "disabled": false,
    },
    "description": undefined,
    "name": "default",
    "table": {
      "category": "slots",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "{ num: unknown }",
      },
    },
    "type": {
      "name": "other",
      "required": false,
      "value": "{ num: unknown }",
    },
  },
  "named": {
    "control": {
      "disabled": false,
    },
    "description": undefined,
    "name": "named",
    "table": {
      "category": "slots",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "{ str: unknown }",
      },
    },
    "type": {
      "name": "other",
      "required": false,
      "value": "{ str: unknown }",
    },
  },
  "no-bind": {
    "control": {
      "disabled": false,
    },
    "description": undefined,
    "name": "no-bind",
    "table": {
      "category": "slots",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": undefined,
    },
    "type": {
      "name": "other",
      "required": false,
      "value": "",
    },
  },
  "vbind": {
    "control": {
      "disabled": false,
    },
    "description": undefined,
    "name": "vbind",
    "table": {
      "category": "slots",
      "defaultValue": null,
      "jsDocTags": undefined,
      "type": {
        "summary": "{ num: unknown; str: unknown }",
      },
    },
    "type": {
      "name": "other",
      "required": false,
      "value": "{ num: unknown; str: unknown }",
    },
  },
}
`;

exports[`extractArgTypes (vue-docgen-api) > should extract slots type for Vue component 1`] = `
{
  "default": {
    "control": {
      "disabled": false,
    },
    "description": "",
    "name": "default",
    "table": {
      "category": "slots",
      "type": {
        "summary": "{ num: number; }",
      },
    },
    "type": {
      "name": "other",
      "value": "{ num: number; }",
    },
  },
  "named": {
    "control": {
      "disabled": false,
    },
    "description": "",
    "name": "named",
    "table": {
      "category": "slots",
      "type": {
        "summary": "{ str: string; }",
      },
    },
    "type": {
      "name": "other",
      "value": "{ str: string; }",
    },
  },
  "no-bind": {
    "control": {
      "disabled": false,
    },
    "description": "",
    "name": "no-bind",
    "table": {
      "category": "slots",
      "type": {
        "summary": "{}",
      },
    },
    "type": {
      "name": "other",
      "value": "{}",
    },
  },
  "vbind": {
    "control": {
      "disabled": false,
    },
    "description": "",
    "name": "vbind",
    "table": {
      "category": "slots",
      "type": {
        "summary": "{ num: number; str: string; }",
      },
    },
    "type": {
      "name": "other",
      "value": "{ num: number; str: string; }",
    },
  },
}
`;
